### ISSUE TITLE
About multipie individuals samples of phasing

### ISSUE THREAD

**Issue by myshu2017-03-14** [PacificBiosciences/targeted-phasing-consensus#4]
Hi, 
I  followed the procedures to test my data. But I woder if the **targeted-phasing-consensus.sh** can only be used for one individual? 
I learnt that **RS_ReadsOfInsert_Mapping** protocol can not split different samples according to the barcodes. And I didn't find any Post-processing in the program for demultiplexing. 
What's more,  I find a paper [http://www.pacb.com/wp-content/uploads/multiplex-target-enrichment-barcoded-multi-kilobase-fragments-probe-based-capture-technologies.pdf](url). which seems to mention **RS_ReadsOfInsert_Mapping** protocol to do the phasing, and  **RS_Resequencing_ Barcode.1** protocol is used for enrichment-fold analysis. But what the relationship between them?
Can you help me?
Thanks very much!
My Shu

**Comment by myshu2017-03-14** [comment]
And I find another program.....[https://github.com/PacificBiosciences/Multiplex-Targeted-Sequencing-for-Probe-based-Enrichment-Capture-Investigations](url)
But it seems not been pulished....

**Comment by rhallPB** [comment]
If I understand you correctly you have a sequencing run that contains barcoded pull down enriched samples from multiple individuals that you would like to phase? 
Do you have access to the latest version of SMRT Link 4.0? If you do you can assign barcodes to your data then run the CCS mapping on a per barcode basis (the same way it is done in this example [https://github.com/PacificBiosciences/apps-scripts/tree/master/multiplexHGAP#manual-hgap-per-barcode-using-gui](https://github.com/PacificBiosciences/apps-scripts/tree/master/multiplexHGAP#manual-hgap-per-barcode-using-gui) followed by the bam style phasing [https://github.com/PacificBiosciences/targeted-phasing-consensus#targeted-sequel-phasingsh---sequel-and-smrt-link-40](https://github.com/PacificBiosciences/targeted-phasing-consensus#targeted-sequel-phasingsh---sequel-and-smrt-link-40).

Using SMRT portal 2.3 it is possible, but not very straightforward.

**Comment by myshu2017-03-14** [comment]
@rhallPB Yes, you are right. Thank you very much! And now I'm tring to install SMRT link 4.0. 

What's more, I learned that all of them do the phasing using the samtools. Do you know what the main mechanism of it?
The code is just as below:
`samtools calmd -AEur subset.bam $REFERENCE | samtools phase -b prefix - > phase.out`

**Comment by rhallPB** [comment]
[https://www.biostars.org/p/91604/](https://www.biostars.org/p/91604/
)

**Comment by myshu2017-03-14** [comment]
@rhallPB Hi, I met some problems to run SMRTlink....
First, I used **convert RS to BAM** to convert RS II datas to BAM datas. And then I used **Barcoding** to get a  barcoded xml file. And then use the `dataset split` to split my samples by barcode. Then using `dataset consolidate`  to convert them to BAM files....Is that right?

What I'm confused is that where the **scraps bam**  file, the input of **targeted-sequel-phasing.sh** ??

I konw that the origin of sequel data includes the **scraps bam** file. But when I convert RS II data to BAM and import it to srmtlink, there isn't a scraps bam file. 
> "Scraps" BAM files are always required to reconstitute the ZMW reads internally. Conversely, "scraps" BAM files will be output

Is it necessary for the targeted-sequel-phasing.sh?

**Comment by williamrowell** [comment]
I have a (slow) workaround.  You’ll need to define ```$BCFINDEX``` (barcode forward index) and ```$BCRINDEX``` (barcode reverse index), and the following code should replace lines 37-48 in ```targeted-sequel-phasing.sh```:

```
	# generate a comma-separated list of all read prefixes from QNAME in CCSBAM
	# for example:
	# QNAME = m54026_161028_224529/4260379/0_3848
	# prefix = m54026_161028_224529/4260379
	echo "generating a list of subreads corresponding to phase $PHASE"
	echo "--------------------------------------------------"
	READLIST=`samtools view phase.$PHASE.bam | awk '{ print $1 }' | cut -f1 | cut -d'/' -f1-2 | xargs -I "%" echo %/ | sort | uniq | paste -sd'|'`

	echo "filtering reads corresponding to phase $PHASE and barcode $BCFINDEX,$BCRINDEX"
	echo "--------------------------------------------------"
	samtools view -h $SUBREADSBAM | egrep '^@|bc:B:S,$BCFINDEX,$BCRINDEX' | egrep '^@|$READLIST' | samtools view -bS - > phase.$PHASE.subreads.bam
```
```$READLIST``` here changes from a list of the ZMWs to a list of the QNAME prefixes.  Then, we use egrep to filter the subreads BAM for:
1) the barcode information (```egrep '^@|bc:B:S,$BCFINDEX,$BCRINDEX'```)
2) the QNAME prefixes from the phased CCS reads. (```egrep '^@|$READLIST'```)

**Comment by myshu2017-03-14** [comment]
@williamrowell What are the $BCFINDEX (barcode forward index) and $BCRINDEX (barcode reverse index)? Can you explain them in detail?

**Comment by myshu2017-03-14** [comment]
@williamrowell And did you mean that you used all samples to run CCS mapping, and then split them after `samtools phase` ?

**Comment by williamrowell** [comment]
@myshu2017-03-14 
How many barcodes did you use, and do you happen to know their names or indexes?  The barcode index is the position of the barcode within the barcode reference fasta file.

I've modified the script and provided it in [this gist](https://gist.github.com/williamrowell/098c8b1ec9b587c20e447686b8dde166).

I assume that you have these files/values available:
1) ```CCSBAM``` - This is the consolidated bam file containing mapped CCS reads for one barcoded sample (created by the "Resequencing" analysis).  If the data are not consolidated into a single bam, you'll need to run ```dataset consolidate```.
2) ```SUBREADSBAM``` - This is the consolidated bam file containing all of the subreads for the lane, with each read tagged with the barcode (created by the "Barcoding" analysis).   If the data are not consolidated into a single bam, you'll need to run ```dataset consolidate```.
3) ```ROINAME``` - This is just some name you want to use for this region of interest.  It's only used to name the folder that will contain the haplotype data.
4) ```CHROM```, ```START```, ```END``` - coordinates for the region of interest.
5) ```REF``` - genome reference genome fasta
6) ```BCFINDEX```, ```BCRINDEX``` - If this is standard barcoding, these will have the same value for each barcoded sample.  For instance, for a sample with the first barcode, the values would both be 0.  (The barcode fasta file is zero-indexed.).  For the second sample, the values would both be 1.

You will need to download the script into a directory with ``faidx.zip`` and run the script once per barcode, changing the values for ```CCSBAM```, ```BCFINDEX```, and ```BCRINDEX``` each time.

I'm working on a faster solution for filtering the reads, but I'm not sure when it will be ready.

**Comment by myshu2017-03-14** [comment]
@williamrowell We have about 10 barcode. And barcode information is determined at the time of sequencing. But I'm still wonder where can I get the information of barrcode index? Do I need to calculate my own?

**Comment by williamrowell** [comment]
@myshu2017-03-14 Did you run the "Barcoding" analysis on the raw subreads?  If so, one way to determine the barcodes used is to go to the "Analysis Results" page for the barcoding analysis and check the "Barcodes" section.  It will list:
- Barcode Name - which is the forward and reverse index for the barcode
- Reads - number of reads assigned to this barcode
- Bases - number of bases assigned to this barcode
You would have already checked this analysis output and found the barcode indices if you ran per-barcode CCS mapping.  If you haven't done this yet, you will need to before you can phase the per-barcode CCS reads.

**Comment by myshu2017-03-14** [comment]
@williamrowell Yes, I have run the "Barcoding" analysis on the raw subreads. But our results just as below:

> **Barcodes
> Barcode Name	Reads	Bases**
> -1---1	1,672	9,051,405
> 0--0	33,152	120,474,852
> 1--1	2,542	10,304,358
> ......

There are "-1---1"...But what the index means? Are there some thing wrong with the "Barcoding" analysis results?

**Comment by williamrowell** [comment]
@myshu2017-03-14 
`-1---1` describes reads for which no barcode was determined.  There's nothing wrong with the "Barcoding" analysis results.
After "Barcoding" completed, did you run a new "Resequencing" job for each barcode determined by the "Barcoding" analysis?

**Comment by myshu2017-03-14** [comment]
@williamrowell Why is the "Resequencing" job? why not the "CCS Mapping"? In the shell script` targeted-sequel-phasing.sh` , we need the CCSBAM. And what the meaning the "Resequencing" job done here?

What's more, after running the "Barcoding" job using 96 barcode fasta, I extracted the first 10 samples of the largest number of reads. But their barcode name can't correspounding to my sample....That's too strange.
So I tried to rerun the "Barccoding" using selected barcode fasta, which were part of whole 96 barcode fasta. But I found that we can't run the "Barcoding" successfully . There were always some errors...Did you meet with the same problem? The error massage just as below:

> /io/dataset/DataSetMembers.py", line 646, in <lambda>
>     'bcf': (lambda x: x.bcForward),
>   File "/opt/pacbio_soft/smrtlink/install/smrtlink-fromsrc_4.0.0.190159+190159-190159-189856-189856-189856/bundles/smrttools/install/smrttools-fromsrc_4.0.0.190159/private/thirdparty/pythonpkgs/numpy/numpy_1.9.2/lib/python2.7/site-packages/numpy/core/records.py", line 417, in __getattribute__
>     raise AttributeError("record array has no attribute %s" % attr)
> AttributeError: record array has no attribute bcForward
>

**Comment by williamrowell** [comment]
OK, so I think we need to take an inventory of what you have.

Which analysis jobs did you run in which order?  Did you run a single "Barcoding" job followed by multiple "CCS Mapping" jobs (one per barcode, or did you run a "CCS with Barcoding" job?  I'm trying to determine which input files you have available?

Also, could you paste the full output of the barcoding analysis on the raw subreads.

If you want to match the barcode *names* to the barcode *indices*, @jrharting developed the short python script [here](https://gist.github.com/williamrowell/7258f8048319877e8da198b7be387e26).  To use it, go to your "Barcoding" job and click on `Barcodes`, then copy the barcodeset xml path from the `additional details` section.  Run the script as `python showBarcodeIndices.py <insert barcode xml path here>`.

**Comment by myshu2017-03-14** [comment]
@williamrowell 
1. First, I run a single "Barcoding" job. And the output was a .xml file. 
2. Next, I used "dataset split" to split the .xml file, which outputs were many "chunk0-95" files. 
3. Then, I upload all the split files to smrtlink and run multiple "CCS Mapping" jobs((one per barcode).

All the output of "Barcoding" analysis just as below:
> **Barcodes**
> Barcode Name	Reads	Bases
> -1---1	1,672	9,051,405
> 0--0	33,152	120,474,852
> 1--1	2,542	10,304,358
> 10--10	12,304	50,507,462
> 11--11	18,104	65,802,724
> 12--12	6,969	27,391,487
> 13--13	1,077	4,612,653
> 14--14	7,990	34,122,256
> 15--15	4,217	17,935,928
> 16--16	14,405	54,421,358
> 17--17	5,243	22,342,132
> 18--18	34,621	126,045,184
> 19--19	16,574	64,547,710
> 2--2	1,269	5,872,398
> 20--20	6,691	25,354,355
> 21--21	41,057	147,762,817
> 22--22	8,796	33,376,721
> 23--23	33,634	111,125,328
> 24--24	2,827	13,025,211
> 25--25	75,481	277,843,434
> 26--26	8,567	36,096,596
> 27--27	2,084	9,350,174
> 28--28	4,613	19,964,800
> 29--29	38,714	134,306,618
> 3--3	3,926	16,088,387
> 30--30	3,195	13,662,555
> 31--31	2,080	8,385,055
> 32--32	27,610	110,471,648
> 33--33	24,024	80,560,198
> 34--34	5,431	20,581,429
> 35--35	2,469	10,366,872
> 36--36	5,126	21,560,088
> 37--37	5,278	21,589,792
> 38--38	3,161	13,520,606
> 39--39	12,243	44,909,276
> 4--4	77,968	260,534,052
> 40--40	3,898	14,092,676
> 41--41	2,987	14,100,041
> 42--42	78,541	248,726,299
> 43--43	6,940	26,810,130
> 44--44	14,903	57,653,126
> 45--45	4,738	19,136,721
> 46--46	9,516	36,826,529
> 47--47	49,683	167,580,746
> 48--48	33,428	120,099,749
> 49--49	8,254	31,711,807
> 5--5	1,440	6,690,489
> 50--50	2,027	9,458,272
> 51--51	50,767	184,550,587
> 52--52	1,251	5,636,527
> 53--53	338	1,588,020
> 54--54	6,732	27,448,932
> 55--55	2,911	12,061,199
> 56--56	19,820	77,001,528
> 57--57	5,610	23,261,523
> 58--58	11,271	45,614,703
> 59--59	18,479	63,367,498
> 6--6	22,464	79,747,033
> 60--60	10,243	40,885,383
> 61--61	5,552	24,074,739
> 62--62	3,556	15,010,883
> 63--63	3,472	14,778,300
> 64--64	5,256	19,832,120
> 65--65	9,732	37,989,297
> 66--66	4,654	18,929,526
> 67--67	8,641	32,931,248
> 68--68	35,057	122,701,125
> 69--69	17,549	68,502,991
> 7--7	5,910	22,926,781
> 70--70	3,933	16,686,590
> 71--71	10,903	40,326,854
> 72--72	108,149	418,999,496
> 73--73	3,866	16,371,356
> 74--74	29,170	108,288,326
> 75--75	4,574	20,569,394
> 76--76	3,315	14,658,076
> 77--77	5,757	23,167,815
> 78--78	4,163	16,375,332
> 79--79	2,401	10,776,892
> 8--8	2,013	9,106,116
> 80--80	102,494	356,588,426
> 81--81	2,880	12,217,311
> 82--82	5,046	19,966,789
> 83--83	58,250	221,534,959
> 84--84	1,293	5,598,915
> 85--85	2,926	12,358,446
> 86--86	4,543	17,690,675
> 87--87	1,837	7,599,576
> 88--88	6,048	25,797,481
> 89--89	8,676	31,558,709
> 9--9	4,705	19,248,668
> 90--90	3,102	12,758,213
> 91--91	1,482	6,323,092
> 92--92	15,662	61,585,875
> 93--93	1,619	7,369,773
> 94--94	2,858	12,754,791
> 95--95	531	2,877,856

What's more, I tried the python script.` showBarcodeIndices.py`. But there was an error.
`$ python showBarcodeIndices.py /opt/pacbio_soft/smrtlink/userdata/jobs_root/000/000114/tasks/pbcoretools.tasks.gather_subreadset-1/file.subreadset.xml `

> Traceback (most recent call last):
>   File "showBarcodeIndices.py", line 8, in <module>
>     bcSet = BarcodeSet(bcXml)
>   File "/opt/pacbio_soft/smrtlink/install/smrtlink-fromsrc_4.0.0.190159+190159-190159-189856-189856-189856/bundles/smrttools/install/smrttools-fromsrc_4.0.0.190159/private/pacbio/pythonpkgs/pbcore/lib/python2.7/site-packages/pbcore/io/dataset/DataSetIO.py", line 4218, in __init__
>     super(BarcodeSet, self).__init__(*files, **kwargs)
>   File "/opt/pacbio_soft/smrtlink/install/smrtlink-fromsrc_4.0.0.190159+190159-190159-189856-189856-189856/bundles/smrttools/install/smrttools-fromsrc_4.0.0.190159/private/pacbio/pythonpkgs/pbcore/lib/python2.7/site-packages/pbcore/io/dataset/DataSetIO.py", line 3692, in __init__
>     super(ContigSet, self).__init__(*files, **kwargs)
>   File "/opt/pacbio_soft/smrtlink/install/smrtlink-fromsrc_4.0.0.190159+190159-190159-189856-189856-189856/bundles/smrttools/install/smrttools-fromsrc_4.0.0.190159/private/pacbio/pythonpkgs/pbcore/lib/python2.7/site-packages/pbcore/io/dataset/DataSetIO.py", line 432, in __init__
>     files[0])
> IOError: [Errno 5] Cannot create PacBio.DataSet.BarcodeSet from PacBio.DataSet.SubreadSet: '/opt/pacbio_soft/smrtlink/userdata/jobs_root/000/000114/tasks/pbcoretools.tasks.gather_subreadset-1/file.subreadset.xml'

**Comment by myshu2017-03-14** [comment]
@williamrowell I also found that the output directory of Barcoding job had 24 barcode directory..

>  pbcoretools.tasks.bam2bam_barcode-1
>  pbcoretools.tasks.bam2bam_barcode-10
>  pbcoretools.tasks.bam2bam_barcode-11
>  pbcoretools.tasks.bam2bam_barcode-12
>  pbcoretools.tasks.bam2bam_barcode-13
>  pbcoretools.tasks.bam2bam_barcode-14
>  pbcoretools.tasks.bam2bam_barcode-15
>  pbcoretools.tasks.bam2bam_barcode-16
>  pbcoretools.tasks.bam2bam_barcode-17
>  pbcoretools.tasks.bam2bam_barcode-18
>  pbcoretools.tasks.bam2bam_barcode-19
>  pbcoretools.tasks.bam2bam_barcode-2
>  pbcoretools.tasks.bam2bam_barcode-20
>  pbcoretools.tasks.bam2bam_barcode-21
>  pbcoretools.tasks.bam2bam_barcode-22
>  pbcoretools.tasks.bam2bam_barcode-23
>  pbcoretools.tasks.bam2bam_barcode-24
>  pbcoretools.tasks.bam2bam_barcode-3
>  pbcoretools.tasks.bam2bam_barcode-4
>  pbcoretools.tasks.bam2bam_barcode-5
>  pbcoretools.tasks.bam2bam_barcode-6
>  pbcoretools.tasks.bam2bam_barcode-7
>  pbcoretools.tasks.bam2bam_barcode-8
>  pbcoretools.tasks.bam2bam_barcode-9
>  pbcoretools.tasks.gather_subreadset-1
>  pbcoretools.tasks.scatter_subreads_bam2bam-1
>  pbreports.tasks.barcode_report-0

What the barcode folders means??

**Comment by williamrowell** [comment]
@myshu2017-03-14 

1. To answer your last question first (**`What the barcode folders means??`**), the barcoding job was split into smaller tasks to complete faster.  Each of those folders represents one of these smaller tasks.

2. Regarding the script to show barcode names and indices, (**`What's more, I tried the python script.showBarcodeIndices.py. But there was an error.`**), you need to provide the xml for the fasta containing the barcode sequence, not the xml for Barcoding result.
  - Open your `Barcoding` job.
  - Look for the `Inputs` section.
  - Click on `Barcode`.
  - On this page, click `Additional Details`.
  - Copy the xml file from the `Path` section.

3. Did you say [earlier](https://github.com/PacificBiosciences/targeted-phasing-consensus/issues/4#issuecomment-293759695) that you expect ~10 samples?  If that's the case, then I'm a little concerned after looking at the full output of the `Barcoding` task that the incorrect barcode list was chosen.  Did you happen to use barcoded SMRTbell adaptors?  If so, then make sure you chose the `pacbio_barcode_adapters_96` list when creating the barcoding analysis.


Let me know the answer to 3, and we can move on to the next part.

**Comment by myshu2017-03-14** [comment]
@williamrowell Thank you very much for your answer. And I run the `showBarcodeIndices.py` successfully. Part of results just as below:

> 0 lbc1
> 1 lbc2
> 2 lbc3
> 3 lbc4
> 4 lbc5
> 5 lbc6
> 6 lbc7
> 7 lbc8
> 8 lbc9
> 9 lbc10
> 10 lbc11
> 11 lbc12
> 12 lbc13
> 13 lbc14
> 14 lbc15
> 15 lbc16
And the first number is the barcode index, right?

For the 3 question, we use the `pacbio_barcodes_96` not `pacbio_barcode_adapters_96`. 

And we need to extract ~10 samples, which barcode name is konwn,  from 96 results. 
The "Barcoding" results have been ranked according to the number of reads. As expected, the top ~10 should correspond to our sample barcode names, but they are not...

**Comment by williamrowell** [comment]
@myshu2017-03-14 You're interpreting the output of the python script correctly.  Each row is the barcode index (reported in the `Barcoding` analysis output) followed by the barcode name.

I was asking earlier if, in the lab, you used barcoded SMRTbell adaptors.  If so, then you might need to demultiplex using a different barcode list (`pacbio_barcode_adapters_96`).  The reason I suspect this might be the case is that in [your `Barcoding` output](https://github.com/PacificBiosciences/targeted-phasing-consensus/issues/4#issuecomment-295538398), there don't appear to be 10 samples that get a clear majority of the reads.  Try demultiplexing with `pacbio_barcode_adapters_96` instead and see if that helps.

**Comment by myshu2017-03-14** [comment]
@williamrowell Yes, you are right! I tried `pacbio_barcode_adapters_96` And result was normal finally. But I don't know why? I have run `RS_Subreads.1 protocol `using the same sequences of  `pacbio_barcode_96` in SMRT protal. The result was right, I can find my ~10 sample's easilly by sorting reads number. 
But why is the `pacbio_barcode_adapters_96`  in SMRTLINK?
I learnt something from the website [ https://github.com/PacificBiosciences/SMRT-Link/wiki/SMRT-Barcoding-Analysis-Primer#symmetric-mode](url) The  `pacbio_barcode_adapters_96` is reverse complementary sequence of  `pacbio_barcode_96`.
But I still don't know the reason... When should I choose the `pacbio_barcode_96` , and when choose  `pacbio_barcode_adapters_96`.

**Comment by williamrowell** [comment]
@myshu2017-03-14 If I understand correctly, the SMRT Portal software would try to detect all orientations of each barcode, while the SMRT Link software requires the user to be more explicit and provide the barcodes in the correct orientation.  When you use barcoded SMRTbell adaptors (as you appeared to) and analyze using SMRT Link, you will always need to use the reverse complementary sequences in `pacbio_barcode_adapters_96`.

**Comment by myshu2017-03-14** [comment]
@williamrowell Thank you very much!

**Comment by myshu2017-03-14** [comment]
@williamrowell I'm sorry to disturb you again...When I run [the script](https://gist.github.com/williamrowell/098c8b1ec9b587c20e447686b8dde166) you provided before, there was an error `/usr/bin/egrep: Argument list too long` And the error was from the code :

> echo "filtering reads corresponding to phase $PHASE and barcode $BCFINDEX,$BCRINDEX"
> echo "--------------------------------------------------"
> samtools view -h $SUBREADSBAM | egrep "^@|bc:B:S,$BCFINDEX,$BCRINDEX" | egrep "^@|$READLIST" | samtools view -bS - > phase.$PHASE.subreads.bam

Note that I change the single quotes into double quotes (egrep '' to egrep "") for there are variables in egrep matching.

I don't know how to fix the  `/usr/bin/egrep: Argument list too long` error .. Can you help me?? Thank you very much!

**Comment by williamrowell** [comment]
@myshu2017-03-14 I'm happy to help.  It's possible that the number of reads made the command exceed the limit for the command line.  To test this, instead of saving the read prefixes as the variable `$READLIST`, I saved them into `whitelist.txt` and used `egrep -f whitelist.txt`.  I've updated [the gist](https://gist.github.com/williamrowell/098c8b1ec9b587c20e447686b8dde166) to include your changes and this `whitelist.txt` change.  Let me know if it works.
